### PR TITLE
[mesheryctl] improve the error messages shown to users for "mesheryctl pattern"

### LINKED ISSUE #6271: [mesheryctl] improve the error messages shown to users for "mesheryctl pattern"
**Description**: Improve the error messages in mesheryctl commands from the user's perspective. 
**Commands in scope of this issue**: `mesheryctl pattern`, `mesheryctl pattern list`, `mesheryctl pattern apply`, `mesheryctl pattern view`, `mesheryctl pattern delete`

---

We need to make sure of the following things
- There are proper error messages when users
  - Give more arguments to a command than required
  - Misspell a command
  - Do not give the required input via flags or args
  - Use the command without any arguments or flags.
- The error messages being displayed to users are meaningful from user's perspective. And if applicable, a potential fix for the problem should be shown too.  For example
  - "failed to unmarshal response body" could be modified to "couldn't process JSON response from server"
  - "manifest not found" could be modified to "manifest not found. Please provide the path to the app manifest."
  
 --- 
  #### Contributor Resources

- ⌨️ [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

For additional context, head over to this [pr](https://github.com/meshery/meshery/pull/5661).

### PR DESCRIPTION
**Pull request by kamalbuilds**
**Description**

This PR fixes #6271 

**Notes for Reviewers**
1> A notable case considered which is resolved in this PR is 
image

`mesheryctl pattern view` assumes a name if ID doesn't match so that case is covered in .

2> Authentification fix via CLI is done.
3>Unmarshall code is reframed from user's perspective.

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: mesheryctl/internal/cli/root/system/error.go [hunk 1]
```diff
@@ -63,7 +63,7 @@ func ErrCreateDir(err error, obj string) error {
 }
 
 func ErrUnmarshal(err error, obj string) error {
-	return errors.New(ErrUnmarshalCode, errors.Alert, []string{"Error unmarshalling file ", obj}, []string{err.Error()}, []string{}, []string{})
+	return errors.New(ErrUnmarshalCode, errors.Alert, []string{"Error processing JSON response from server to file", obj}, []string{err.Error()}, []string{}, []string{})
```
Comment (hershd23) [inline mesheryctl/internal/cli/root/system/error.go:66]: `processing JSON response from server to file`

Seems a little weird if we aren't unmarshalling files. Is this only handling files? If not just omit `to file`

File: mesheryctl/pkg/utils/error.go [hunk 2]
```diff
@@ -128,7 +128,7 @@ func formatError(msg string, cmd cmdType) string {
 
 func ErrAttachAuthToken(err error) error {
 	return errors.New(ErrAttachAuthTokenCode, errors.Alert, []string{err.Error()},
-		[]string{"Authentication token not found. please supply a valid user token with the --token (or -t) flag. or login with `mesheryctl system login`"}, []string{}, []string{})
+		[]string{"Authentication token not found. please supply a valid user token using `mesheryctl system config [aks|eks|gke|minikube] -t <path>`. or login with `mesheryctl system login`"}, []string{}, []string{})
```
Comment (hershd23) [inline mesheryctl/pkg/utils/error.go:131]: I think we have the token flag for quite a few methods. Have we changed that? This seems a generic error type, not specific to `mesheryctl system config`

Also not sure if we should delete the `--token (or -t)` suggestion

**Review by hershd23 — COMMENTED** [review]
A couple of comments. Let me know if my understanding about some of these is incorrect

File: mesheryctl/pkg/utils/error.go (hunk 2 above)
Comment (legendarykamal) [reply mesheryctl/pkg/utils/error.go:131]: the idea here was not to delete the --token flag but to provide the full command i.e. `mesheryctl system config  [aks|eks|gke|minikube] -t <path>` which inturn will help the user in providing the token via this command. 

so the --token suggestion is not deleted but is intact to the helping command.

Comment (hershd23) [reply mesheryctl/pkg/utils/error.go:131]: So your command is an example right? Like this auth error comes on other commands as well right. Is it right to just specify one example?

Comment (legendarykamal) [reply mesheryctl/pkg/utils/error.go:131]: An example or a case, the case here considered is 
user is trying to list the pattern using `mesheryctl pattern list` but he is not authenticated. 
So a token is being asked from him to supply via the `mesheryctl system config [aks|eks|gke|minikube] -t <path>` cmd  or login with `mesheryctl system login` .

Comment (hershd23) [reply mesheryctl/pkg/utils/error.go:131]: I don't think a specific case should be mentioned for a generic authentication error type. 

@revolyssup

File: mesheryctl/internal/cli/root/system/error.go (hunk 1 above)
Comment (legendarykamal) [reply mesheryctl/internal/cli/root/system/error.go]: done 👍

File: mesheryctl/pkg/utils/error.go (hunk 2 above)
Comment (legendarykamal) [reply mesheryctl/pkg/utils/error.go:131]: Ok @alphaX86 @hershd23 thank you for the feedback, I will change this accordingly 🙂 no sweat.

**Comment by legendarykamal** [comment]
@nishantwrp could you please review this PR and tell If any modification are needed or this is good to go? 

@hershd23 review has been addressed succerssfully.

**Comment by hershd23** [comment]
For most part changes suggested here and in #6356 replace `unmarshall` with couldn't process JSON. 

@legendarykamal @XDRAGON2002 if you can decide on one string that would be better. @alphaX86 do you have any ideas/what the default error string for unmarshalling should be?

**Comment by alphaX86** [comment]
> For most part changes suggested here and in #6356 replace `unmarshall` with couldn't process JSON.
> 
> @legendarykamal @XDRAGON2002 if you can decide on one string that would be better. @alphaX86 do you have any ideas/what the default error string for unmarshalling should be?
> 
> EDIT:- Actually they are making changes to the same files. Can we standardize the error messages?
> 
> Otherwise I don't have any issues with this. Once this is resolved, i'll approve

Well, for unmarshal, either we can mention as can't process json data or couldn't convert byte to json. It depends on our way of converting error to users

**Comment by legendarykamal** [comment]
@alphaX86 @hershd23  both of us are using can't process json data , should we replace it with couldn't convert byte to json. 

PS:- First approach sounds better than the 2nd in my opinion .

File: mesheryctl/internal/cli/root/pattern/view.go [hunk 3]
```diff
@@ -106,7 +106,7 @@ mesheryctl pattern view [pattern-name | ID]
 			// use the first match from the result when searching by pattern name
 			arr := dat["patterns"].([]interface{})
 			if len(arr) == 0 {
-				utils.Log.Info(fmt.Sprintf("pattern with name: %s not found", pattern))
+				utils.Log.Info(fmt.Sprintf("pattern with name: %s not found , please enter the correct [pattern name | ID] ", pattern))
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/pattern/view.go:109]:
```suggestion
				utils.Log.Info(fmt.Sprintf("pattern with name: %s not found. Enter a valid pattern name or ID", pattern))
```

File: mesheryctl/internal/cli/root/system/error.go [hunk 4]
```diff
@@ -63,7 +63,7 @@ func ErrCreateDir(err error, obj string) error {
 }
 
 func ErrUnmarshal(err error, obj string) error {
-	return errors.New(ErrUnmarshalCode, errors.Alert, []string{"Error unmarshalling file ", obj}, []string{err.Error()}, []string{}, []string{})
+	return errors.New(ErrUnmarshalCode, errors.Alert, []string{"Error processing JSON response from server", obj}, []string{err.Error()}, []string{}, []string{"Either the JSON response or the Response is distorted"})
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/error.go:66]:
```suggestion
	return errors.New(ErrUnmarshalCode, errors.Alert, []string{"Error processing JSON response from Meshery Server", obj}, []string{err.Error()}, []string{}, []string{"Either the JSON response or the Response is distorted"})
```

File: mesheryctl/internal/cli/root/pattern/view.go [hunk 5]
```diff
@@ -90,7 +90,7 @@ mesheryctl pattern view [pattern-name | ID]
 
 		var dat map[string]interface{}
 		if err = json.Unmarshal(body, &dat); err != nil {
-			return errors.Wrap(err, "failed to unmarshal response body")
+			return errors.Wrap(err, "couldn't process JSON response from server")
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/pattern/view.go:93]:
```suggestion
			return errors.Wrap(err, "couldn't process JSON response from Meshery Server")
```

File: mesheryctl/internal/cli/root/pattern/view.go [hunk 6]
```diff
@@ -57,7 +57,7 @@ mesheryctl pattern view [pattern-name | ID]
 			if viewAllFlag {
 				url += "/api/pattern?page_size=10000"
 			} else {
-				return errors.New("[pattern-name|pattern-id] not specified, use -a to view all patterns")
+				return errors.New("[pattern-name|pattern-id] is not specified, use -a to view all patterns")
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/pattern/view.go:60]:
```suggestion
				return errors.New("Pattern name or ID is not specified. Use `-a` to view all patterns")
```

File: mesheryctl/pkg/utils/error.go [hunk 7]
```diff
@@ -128,7 +128,7 @@ func formatError(msg string, cmd cmdType) string {
 
 func ErrAttachAuthToken(err error) error {
 	return errors.New(ErrAttachAuthTokenCode, errors.Alert, []string{err.Error()},
-		[]string{"Authentication token not found. please supply a valid user token with the --token (or -t) flag. or login with `mesheryctl system login`"}, []string{}, []string{})
+		[]string{"Authentication token not found. Please supply a valid user token using the --token (or -t) flag. Or login with `mesheryctl system login`"}, []string{}, []string{})
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/error.go:131]:
```suggestion
		[]string{"Authentication token not found. Login with `mesheryctl system login` or supply a valid user token using the --token (or -t) flag."}, []string{}, []string{})
```

File: mesheryctl/internal/cli/root/pattern/delete.go [hunk 8]
```diff
@@ -42,7 +42,7 @@ mesheryctl pattern delete [file | URL]
 		if !govalidator.IsURL(file) {
 			content, err := os.ReadFile(file)
 			if err != nil {
-				return errors.New(utils.PatternError(fmt.Sprintf("failed to read file %s", file)))
+				return errors.New(utils.PatternError(fmt.Sprintf("failed to read file %s Ensure the file name or URL is correct", file)))
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/pattern/delete.go:45]:
```suggestion
				return errors.New(utils.PatternError(fmt.Sprintf("failed to read file %s. Ensure the filename or URL is valid", file)))
```

**Review by leecalcote — APPROVED** [review]

**Pull request merged**